File: android/src/main/kotlin/com/cloudflare/realtimekit/flutter/view/videoview/RtkVideoViewFactory.kt
package com.cloudflare.realtimekit.flutter.view.videoview

import android.content.Context
import com.cloudflare.realtimekit.RealtimeKitClient
import com.cloudflare.realtimekit.flutter.RtkClientProvider
import io.flutter.plugin.common.StandardMessageCodec
import io.flutter.plugin.platform.PlatformView
import io.flutter.plugin.platform.PlatformViewFactory

class RtkVideoViewFactory(
    private val mobileClient: RealtimeKitClient,
    ) :
    PlatformViewFactory(StandardMessageCodec.INSTANCE) {

    override fun create(context: Context?, viewId: Int, args: Any?): PlatformView {
        // Use the live client (rebuilt per meeting), not the one captured at
        // factory-registration time, so views created after a rejoin resolve
        // participants from the current meeting.
        val mobileClient = RtkClientProvider.requireRealtimeClient()
        val map = args as Map<String, Any?>
        val id = map["id"] as String?
        val isSelfParticipant = map["isSelfParticipant"] as Boolean
        val participant =
            if (!isSelfParticipant && id != null)
                mobileClient.participants.joined.find { it.id == id }!!
            else
                mobileClient.localUser
        if (context == null) {
            throw AssertionError("context for DyteVideoView is null!")
        }

        return RtkVideoPlatformView(context, participant)

    }

}

File: ios/realtimekit_core_ios/Sources/realtimekit_core_ios/view/RtkVideoViewFactory.swift
import Flutter
import RealtimeKitFlutterCoreKMM

class RtkVideoViewFactory: NSObject, FlutterPlatformViewFactory {
    private var messenger: FlutterBinaryMessenger
    // Read the live client dynamically so views created after a rejoin (which
    // rebuilds the native client) bind to the current client, not a stale one.
    var mobileClient: CoreRealtimeKitClient { RtkClientProvider.shared.core }

    init(messenger: FlutterBinaryMessenger, client _: CoreRealtimeKitClient) {
        self.messenger = messenger
        super.init()
    }

    func createArgsCodec() -> FlutterMessageCodec & NSObjectProtocol {
        return FlutterStandardMessageCodec.sharedInstance()
    }

    func create(withFrame frame: CGRect, viewIdentifier viewId: Int64, arguments args: Any?) -> FlutterPlatformView {
        guard let map = args as? [String: Any?],
              let id = map["id"] as? String?,
              let isSelfParticipant = map["isSelfParticipant"] as? Bool
        else {
            fatalError("Invalid arguments provided")
        }
        let participant: CoreRtkMeetingParticipant
        if !isSelfParticipant && id != nil {
            participant = mobileClient.participants.joined.first(where: { $0.id == id })!
        } else {
            participant = mobileClient.localUser
        }
        return RtkVideoView(frame: frame, viewIdentifier: viewId, participant: participant)
    }
}
